README.md
# VM
A Virtual Machine and assembler for my first custom instruction set. and I know, it has the most original name ever 🙃.<br>
This is my first attempt at designing an instruction set, so a lot of instructions could be removed and others do their job.<br>
for example `GET <reg>` could be replaced with `PUSH <reg>`.

## Challenge
I challenge anyone who wants to create the most complicated program possible.<br>
You can submit your program(s) [here](https://github.com/Itai-Nelken/vm/issues/new).
### Leaderboard
|        |                    USERNAME                    |               PROGRAM               |          DESCRIPTION            |
| :----: |                    :------:                    |               :-----:               |          :---------:            |
| 1st 🥇 | [@Itai-Nelken](https://github.com/Itai-Nelken) | `examples/multiplication_table.vas` | prints the multiplication table |
| 2nd 🥈 | [@Itai-Nelken](https://github.com/Itai-Nelken) | `examples/fibonacci.vas`            | prints the first 40 numbers in the fibonacci sequence. |
| 3rd 🥉 |                                                |                                     |                                 |

## Building
In the root of the repo run:
```shell
make
```
The executables will be in `build/`.<br>
You can compress the build, examples and syntax folders along with the README.md and CHANGELOG.md files into a `.tar.xz` archive with `make release`.<br>
you can override the following variables when running `make`:
- `CC` - The C compiler to use, default is `gcc`.
- `CFLAGS` - Flags for the C compiler. default is `-Wall -Wextra -Wpedantic` and the include path.
- `OS` - The OS name to use when compressing the build folder, default is the output of `uname -s`.
- `ARCH` - The acrhitecture to use when compressing the build folder, default is the output of `uname -m`.

and provide the following arguments:
- `DEBUG` - Build with debug symbols, default is 0 (off).
- `VERBOSE` - print the full commands being run, default is 0 (off).

## TODO
- [x] Write the actual VM
   * [x] Use a dynamic array for storing the program (instead of the currently used linked list) that way we can use the `pc` register to actually point to the index of an instruction in the program.
- [ ] Create syntax highlighting for:
   * [x] VIM
   * [ ] VSCode
   * [ ] Sublime Text
- [x] Fix register highlighting for vim syntax highlighting.

## Syntax highlighing
- **VIM**<br>
    Copy the files in `syntax/VIM` to `~/.vim`.
- **VSCode**
    * No official extension yet.
    * [This](https://marketplace.visualstudio.com/items?itemName=dan-c-underwood.arm) one for ARM assembly works fine for now.
- **Sublime Text**
    * No official extension yet.

## Instruction table
| NAME | OPCODE | USAGE | DESCRIPTION |
| :---: | :---: | :---: | :---: |
| PUSH | 0 | `PUSH <value>` | Push <value> to the stack. |
| POP | 1 | `POP` | Pop the stack head. |
| PEEK | 2 | `PEEK` | Print the stack head. |
| ADD | 3 | `ADD` | Pop 2 values from the stack, add them, and push the result. |
| SUB | 4 | `SUB` | Pop 2 values from the stack, substract the LAST PUSHED from the FIRST PUSHED, and push the result. (`PUSH 5, PUSH 2, SUB` == 5 - 2) |
| MUL | 5 | `MUL` | Pop 2 values from the stack, multiply them, and push the result. |
| DIV | 6 | `DIV` | Pop 2 values from the stack, divide the FIRST PUSHED by the LAST PUSHED, and push the result. (`PUSH 4, PUSH 2, DIV` == 4 / 2) |
| SET | 7 | `SET <register> <value>` | Set `<register>` to `<value>`. |
| PSET | 8 | `PSET <register>` | Pop the stack head and store the value in `<register>` |
| GET | 9 | `GET <register>` | Push the value of `<register>` to the stack. |
| JMP | 10 | `JMP <pc value>` | Jump to the instruction in `<pc value>`. |
| JEQ | 11 | `JEQ <pc value>` | Pop 2 values of the stack, compare them, and jump to `<pc value>` if they are equal. |
| JNE | 12 | `JNE <pc value>` | Pop 2 values of the stack, compare them, and jump to `<pc value>` if they are NOT equal. |
| JLT | 13 | `JLT <pc value>` | Pop 2 values of the stack, compare them, and jump to `<pc value>` if the FIRST value pushed to the stack is SMALLER than the second. |
| JGT | 14 | `JGT <pc value>` | Pop 2 values of the stack, compare them, and jump to `<pc value>` if the FIRST value pushed to the stack is LARGER than the second. |
| MOV | 15 | `MOV <reg1> <reg2>` | Move the value in `<reg2>` to `<reg1>`. |
| NOP | 16 | `NOP` | Do nothing. |
| HLT | 17 | `HLT` | End the program. |

## Registers
| NAME | number | DESCRIPTION |
| :--: | :----: | :---------: |
| `A` | 0 | General purpose register. |
| `B` | 1 | General purpose register. |
| `C` | 2 | General purpose register. |
| `D` | 3 | General purpose register. |
| `E` | 4 | General purpose register. |
| `F` | 5 | General purpose register. |
| `SP` | 6 | Stack Pointer (points to the stack head). |
| `PC` | 7 | Program Counter/Instruction Pointer (Points to the instruction to execute in the program). |

### Runtime errors
The following conditions will cause a runtime error:
* Stack is full and `PUSH` is executed.
* Stack is empty and `POP` or `PEEK` are executed.
* Division by zero.
* Unknown instruction.

### How to read disassembler output (`vas -d <vm binary>`)
The output of the disassembler contains 3 rows:
1) The instruction number in the program followed by a colon.
2) The opcode number.
3) The opcode + operands.
For example the following line means that instruction `0` in the program is `0x0`, `PUSH` with the operand `10`:
```
    0: 0x0 PUSH 10
```

## binary format
**SYMOBLS:**<br>
`START`: 0xFFFF (-1)<br>
`END`: 0xFEFF (-2)

**INSTRUCTION FORMAT:**<br>
| instruction | operand 1 | operand 2 |
| :---------: | :-------: | :-------: |
|   2 bytes   |  4 bytes  |  4 bytes  |

Internally the opcode + operands are a `struct`:
```c
struct operation {
    short instruction;
    int arg1, arg2;
};
```

C program to write the following VM program:
```asm
push 5
push 6
add
peek
```
```c
#include <stdio.h>

#define S_START -1
#define S_END -2

enum instructions {
    PUSH,
    POP,
    PEEK,
    ADD
}; // only the first four instructions.

struct operation {
    short instruction;
    int arg1, arg2;
};

int main(void) {
    struct operation op;
    FILE *out=fopen("program.vm", "w");
    if(out == NULL) return 1; // return 1 if failed to open file.

    op.instruction = S_START;
    op.arg1 = 0;
    op.arg2 = 0;
    fwrite(&op, sizeof(op), 1, out); // sizeof(op) is the same as sizeof(struct operation)

    op.instruction = PUSH;
    op.arg1 = 5;
    fwrite(&op, sizeof(op), 1, out);

    op.arg1 = 6;
    fwrite(&op, sizeof(op), 1, out);
    op.arg1 = 0;

    op.instruction = ADD;
    fwrite(&op, sizeof(op), 1, out);

    op.instruction = PEEK;
    fwrite(&op, sizeof(op), 1, out);

    op.instruction = S_END;
    fwrite(&op, sizeof(op), 1, out);

    fclose(out);
    return 0;
}
```

## Examples
**For other examples, see the `examples/` folder.**
### (5 + 2) * (2 + 5)
```asm
push 5
push 2
add ; add pops 2 values off the stack, adds them, and pushes the reuslt
pset Ra ; pset pops 1 value of the stack and stores it in the specified register

push 2
push 5
add
; The result is already on the stack, so no need to save it.

get Ra ; get pushes the value in the specified register and pushes it to the stack
mul

peek ; peek simply prints the stack head
pop ; pop isn't necessary here as its the end of the program
```
### count to 10
```asm
set Ra 0 ; set Ra to 0, this isn't neccesary as all registers are zeroed at VM start 

; increment Ra
get Ra
push 1
add
peek ; print the incremented value
pset Ra

; jump back to the 'increment Ra' part if the value in Ra doesn't equal 10
get Ra
push 10
jne 1
hlt ; isn't neccesary, prints 'program halted'
```


CHANGELOG.md
# VM changelog
**DD/MM/YY**

### 18/09/2021
- Initial idea and VM.
- git repo created.
- created README.md and CHAGNELOG.md (this changelog).
- Wrote dissasembler.
- Added command line arguments.

### 19-20/09/2021
- Wrote a very simple assembler.

### 21/09/2021
- Changed binary format to real binary.
- Wrote a new disassembler.
- Started writing a proper assembler.

### 22/09/2021
- Finish the new assembler.
- Start work on the actual VM.

### 23/09/2021
- Improve disassembler: handle failed to open file error and print `R<reg>` instead of `<reg>`.

### 24/09/2021
- VM: implement loading the binary into a linked list (bad idea as I can't iterate over it with the program counter easily).
- Update documentation.

### 27/09/2021
- Assembler: fix incorrect line being printed in syntax error in some cases.
- Update documentation.

### 01/10/2021
- Finished the VM!
- Cleanup a bit
- Fix JMP commands.
- Add example program that prints the multiplication table.
- Added base Makefile.

### 02/10/2021
- Fix typo in VM/src/vm.c

### 03/10/2021
- Fix VIM syntax highlighting
- Clean up syntax/ folder
- Make the disassembler (`vas -d <vm binary>`) print the instruction number in the program before each instruction.

### 08/10/2021
- Add example that prints the first 40 numbers in the fibonacci sequence.

### 15-16/10/2021
- Rewrite the assembler, nothing substantial. just making the code more readable and loading the program into a dynamic array before writing it.

### 24/10/2021
- fix compilation on MacOS.

### 15/11/2021
- Add make VM print debug info (stack and register dump, current instruction) and exit when a runtime error is encountered.

### 16/11/2021
- VM: runtime error on unknown opcode.

### 19/11/2021
- Finally fix the makefiles.
- Refactor assembler and disassembler.

<hr>


src/cpu.c
#include <stdio.h>
#include <stdlib.h>
#include "cpu.h"
#include "common.h"

// instruction names for instr2str()
const char *instrutions[] = {
	"PUSH",  // 0
	"POP",   // 1
	"PEEK",  // 2
	"ADD",   // 3
	"SUB",   // 4
	"MUL",   // 5
	"DIV",   // 6
	"SET",   // 7
	"PSET",  // 8
	"GET",   // 9
	"JMP",   // 10
	"MOV",   // 11
	"NOP",   // 12
	"HLT",   // 13
	"END",   // 14
	"UNKOWN" // 15
};
const char *instr2str(Instruction i) {
	switch(i) {
		case PUSH:
			return instrutions[0];
		case POP:
			return instrutions[1];
		case PEEK:
			return instrutions[2];
		case ADD:
			return instrutions[3];
		case SUB:
			return instrutions[4];
		case MUL:
			return instrutions[5];
		case DIV:
			return instrutions[6];
		case SET:
			return instrutions[7];
		case PSET:
			return instrutions[8];
		case GET:
			return instrutions[9];
		case JMP:
			return instrutions[10];
		case MOV:
			return instrutions[11];
		case NOP:
			return instrutions[12];
		case HLT:
			return instrutions[13];
		case END:
			return instrutions[14];
		default:
			return instrutions[15];
	}
}

const char *registers[] = {
	"A",     // 0
	"B",     // 1
	"C",     // 2
	"D",     // 3
	"E",     // 4
	"F",     // 5
	"PC",    // 6
	"SP",    // 7
	"UNKNOWN" // 8
};
const char *reg2str(Register r) {
	switch(r) {
		case A:
			return registers[0];
		case B:
			return registers[1];
		case C:
			return registers[2];
		case D:
			return registers[3];
		case E:
			return registers[4];
		case F:
			return registers[5];
		case PC:
			return registers[6];
		case SP:
			return registers[7];
		default:
			return registers[8];
	}
}

void exec(EMU_context *c, Instruction i) {
	// make pointers to the SP and PC registers (makes the code look cleaner).
	int *sp=&(c->registers[SP]);
	int *pc=&(c->registers[PC]);
	switch(i) {
		case PUSH: // PUSH <value>
			c->stack[++*sp]=c->program[++*pc];
			break;
		case POP: // DOESN'T print the value.
			--*sp;
			break;
		case PEEK: // prints the stack head.
			printf("%d\n", c->stack[*sp]);
			break;
		case ADD: { // make a scope so variables can be defined
			int a, b;
			a=c->stack[(*sp)--];
			b=c->stack[(*sp)--];
			c->stack[++*sp]=a+b;
			break;
		}
		case SUB: {
			int a, b;
			a=c->stack[(*sp)--];
			b=c->stack[(*sp)--];
			c->stack[++*sp]=b-a;
			break;
		}
		case MUL: {
			int a, b;
			a=c->stack[(*sp)--];
			b=c->stack[(*sp)--];
			c->stack[++*sp]=a*b;
			break;
		}
		case DIV: {
			int a, b;
			a=c->stack[(*sp)--];
			b=c->stack[(*sp)--];
			c->stack[++*sp]=b/a;
			break;
		}
		case SET: { // SET <register> <value>
			int reg=c->program[++*pc];
			c->registers[reg]=c->program[++*pc];
			break;
		}
		case PSET:
			c->registers[c->program[++*pc]]=c->stack[(*sp)--];
			break;
		case GET: // GET <register> (push the value to the stack)
			c->stack[++*sp]=c->registers[c->program[++*pc]];
			break;
		case JMP: { // JMP <pc value>
			int val=c->program[++*pc]-1; // has to be 1 less than what is requested because loop in emu.c -> run() increments it.
			*pc=val;
			break;
		}
		case MOV: {
			Register dest=c->program[++*pc];
			Register src=c->program[++*pc];
			c->registers[dest]=c->registers[src];
			break;
		}
		case NOP: // do nothing
			break;
		case HLT: // stop the program
		case END: // does the same as HLT
			c->isRunning=0;
			break;
		default:
			// unknkown instruction
			fprintf(stderr, "ERROR: unkown instruction '%d'!\npc=%d\nAborted execution\n", i, *pc);
			c->isRunning=0;
			break;
	}
}
